lib/ProductOpener/ContributionKnowledgePanels.pm

```diff
@@ -0,0 +1,158 @@
+
+=head4 country code $target_cc
+
+We may display country specific recommendations from health authorities, or country specific scores.
+
+=head4 options reference $options_ref
+
+=cut
+
+sub create_contribution_card_panel ($product_ref, $target_lc, $target_cc, $options_ref) {
+
+	$log->debug("create contribution card panel", {code => $product_ref->{code}}) if $log->is_debug();
+
+	my @panels = ();
+	for my $tag_type (qw(data_quality_errors data_quality_warnings data_quality_info)) {
+		# we need to create it first because it can condition contribution panel display
+		my $created = create_data_quality_panel($tag_type, $product_ref, $target_lc, $target_cc, $options_ref);
+		push(@panels, $created) if $created;
+	}
+	if (@panels) {
+		my $panel_data_ref = {quality_panels => \@panels,};
+		create_panel_from_json_template("contribution_card",
+			"api/knowledge-panels/contribution/contribution_card.tt.json",
+			$panel_data_ref, $product_ref, $target_lc, $target_cc, $options_ref);
+	}
+	return;
+}
+
+# private sub to add tag names to data_quality_tags_by_actions: we just iterate to add tag name
+sub _add_quality_tags_names ($tagtype, $target_lc, $quality_tags_by_action) {
+	while (my ($action, $tags) = each %$quality_tags_by_action) {
+		while (my ($tagid, $infos) = each %$tags) {
+			$infos->{"display_name"} = display_taxonomy_tag($target_lc, $tagtype, $tagid);
+		}
+	}
+	return;
+}
+
+=head2 create_data_quality_panel ( $tags_type, $product_ref, $target_lc, $target_cc, $options_ref )
+
+Creates knowledge panels to describe quality issues and invite to contribute
+
+=head3 Arguments
+
+=head4 $tags_type - str
+
+The type of tag. Eg. "data_quality_errors"
+
+=head4 product reference $product_ref
+
+Loaded from the MongoDB database, Storable files, or the OFF API.
+
+=head4 language code $target_lc
+
+Returned attributes contain both data and strings intended to be displayed to users.
+This parameter sets the desired language for the user facing strings.
+
+=head4 country code $target_cc
+
+=cut
+
+sub create_data_quality_panel ($tags_type, $product_ref, $target_lc, $target_cc, $options_ref) {
+
+	$log->debug("create quality errors panel", {code => $product_ref->{code}}) if $log->is_debug();
+
+	my $field_name = $tags_type . "_tags";
+	my @data_quality_tags = @{$product_ref->{$field_name} // []};
+	my $created = undef;
+	# Only display to login user on the web !
+	if (   $options_ref->{user_logged_in}
+		&& ($options_ref->{knowledge_panels_client} eq 'web')
+		&& (scalar @data_quality_tags))
+	{
+		my $panel_data_ref = {};
+		my $quality_tags_by_action = tags_by_prop("data_quality", $product_ref->{$field_name},
+			"fix_action:en", ["description:en"], ["show_to:en"]);
+		if (%$quality_tags_by_action) {
+			_add_quality_tags_names($tags_type, $target_lc, $quality_tags_by_action);
+			$panel_data_ref->{tags_type} = $tags_type;
+			$panel_data_ref->{quality_tags} = $quality_tags_by_action;
```

Comment:
It's a bit strange to have $panel_data_ref->{quality_tags} that contains in fact actions. At a minimum, I think we should keep the name in the panel $panel_data_ref->{quality_tags_by_action}, but in fact I think it would be better to reverse and put actions first in the name. $panel_data_ref->{quality_actions}

Reply:
done.